File: backend/parent.py
"""
Parent Dashboard routes
Allows parents to view child's progress, quiz attempts, doubts
"""

from flask import Blueprint, request, jsonify
from flask_login import current_user
from extensions import db
from backend.models import User, StudentProgress, QuizAttempt, Doubt, Quiz
from backend.utils import require_login

parent_bp = Blueprint('parent', __name__, url_prefix='/api/parent')


@parent_bp.route('/children', methods=['GET'])
@require_login
def list_children():
    """Get list of children linked to this parent"""
    if current_user.role != 'parent':
        return jsonify({'error': 'Parent access required'}), 403

    # For now, parents can search by student email
    # In production, you'd link via parent_id
    return jsonify({
        'message': 'Search children by email to view their progress',
        'parent': current_user.full_name
    }), 200


@parent_bp.route('/child/<string:email>/progress', methods=['GET'])
@require_login
def child_progress(email):
    """Get progress of a specific child by email"""
    if current_user.role != 'parent':
        return jsonify({'error': 'Parent access required'}), 403

    child = User.query.filter_by(email=email, role='student').first()
    if not child:
        return jsonify({'error': 'Student not found with this email'}), 404

    # Get progress records
    progress_records = StudentProgress.query.filter_by(student_id=child.id).all()

    total_quizzes = sum(p.quizzes_completed or 0 for p in progress_records)
    total_correct = sum(p.correct_answers or 0 for p in progress_records)
    total_answered = sum(p.total_questions_answered or 0 for p in progress_records)
    overall_accuracy = (total_correct / total_answered * 100) if total_answered > 0 else 0

    # Recent quiz attempts
    attempts = QuizAttempt.query.filter_by(student_id=child.id).order_by(QuizAttempt.started_at.desc()).limit(10).all()

    # Doubts
    open_doubts = Doubt.query.filter_by(student_id=child.id, status='open').count()
    resolved_doubts = Doubt.query.filter_by(student_id=child.id, status='resolved').count()

    return jsonify({
        'child': {
            'id': child.id,
            'name': child.full_name,
            'email': child.email,
            'grade': child.grade_level
        },
        'overall': {
            'quizzes_completed': total_quizzes,
            'total_questions': total_answered,
            'correct_answers': total_correct,
            'accuracy': round(overall_accuracy, 1),
            'open_doubts': open_doubts,
            'resolved_doubts': resolved_doubts
        },
        'by_subject': [{
            'subject': p.subject,
            'quizzes': p.quizzes_completed or 0,
            'questions': p.total_questions_answered or 0,
            'correct': p.correct_answers or 0,
            'accuracy': round((p.correct_answers or 0) / (p.total_questions_answered or 1) * 100, 1),
            'last_activity': p.last_activity.isoformat() if p.last_activity else None
        } for p in progress_records],
        'recent_attempts': [{
            'quiz_title': a.quiz.title if a.quiz else 'Unknown',
            'subject': a.quiz.subject if a.quiz else 'Unknown',
            'percentage': a.percentage,
            'status': a.status,
            'date': a.submitted_at.isoformat() if a.submitted_at else (a.started_at.isoformat() if a.started_at else None)
        } for a in attempts]
    }), 200


@parent_bp.route('/child/<string:email>/doubts', methods=['GET'])
@require_login
def child_doubts(email):
    """Get doubts of a specific child"""
    if current_user.role != 'parent':
        return jsonify({'error': 'Parent access required'}), 403

    child = User.query.filter_by(email=email, role='student').first()
    if not child:
        return jsonify({'error': 'Student not found'}), 404

    doubts = Doubt.query.filter_by(student_id=child.id).order_by(Doubt.created_at.desc()).all()

    return jsonify([{
        'id': d.id,
        'question': d.question_text,
        'subject': d.subject,
        'status': d.status,
        'created_at': d.created_at.isoformat(),
        'response_count': len(d.responses)
    } for d in doubts]), 200

File: frontend/js/parent.js
/**
 * Parent Dashboard
 */

let currentChildEmail = null;

async function loadParentDashboard() {
    const user = Utils.getCurrentUser();
    if (!user || user.role !== 'parent') return;

    const container = document.getElementById('parentContent');
    if (!container) return;

    container.innerHTML = `
        <div class="parent-welcome">
            <h2>👨‍👩‍👧 Parent Dashboard</h2>
            <p>Track your child's learning progress</p>
        </div>

        <div class="parent-search-card">
            <h3>🔍 Find Your Child</h3>
            <p>Enter your child's registered email to view their progress</p>
            <div class="parent-search-form">
                <input type="email" id="childEmail" placeholder="child@example.com">
                <button onclick="searchChild()" class="btn btn-primary">View Progress</button>
            </div>
            <p class="hint-text">💡 Try: student@example.com (demo account)</p>
        </div>

        <div id="childProgressView"></div>
    `;
}

async function searchChild() {
    const email = document.getElementById('childEmail').value.trim();
    if (!email) {
        Utils.showSuccess('Please enter your child\'s email');
        return;
    }

    try {
        const response = await API.request('GET', `/parent/child/${encodeURIComponent(email)}/progress`);
        currentChildEmail = email;
        displayChildProgress(response);
    } catch (error) {
        console.error('Failed to load child:', error);
        document.getElementById('childProgressView').innerHTML = `
            <div class="error-message">
                <p>❌ Could not find student with email: ${email}</p>
                <p>Make sure your child has registered with this email.</p>
            </div>
        `;
    }
}

function displayChildProgress(data) {
    const container = document.getElementById('childProgressView');
    const child = data.child;
    const overall = data.overall;

    let html = `
        <div class="child-info-card">
            <div class="child-avatar">👨‍🎓</div>
            <div>
                <h3>${child.name}</h3>
                <p>Grade ${child.grade} • ${child.email}</p>
            </div>
        </div>

        <h3 class="section-title">📊 Overall Performance</h3>
        <div class="stats-grid">
            <div class="stat-card math">
                <div class="stat-card-icon">📝</div>
                <div class="stat-card-number">${overall.quizzes_completed}</div>
                <div class="stat-card-label">Quizzes Completed</div>
            </div>
            <div class="stat-card science">
                <div class="stat-card-icon">✅</div>
                <div class="stat-card-number">${overall.accuracy}%</div>
                <div class="stat-card-label">Accuracy</div>
            </div>
            <div class="stat-card english">
                <div class="stat-card-icon">📚</div>
                <div class="stat-card-number">${overall.total_questions}</div>
                <div class="stat-card-label">Questions Answered</div>
            </div>
            <div class="stat-card social">
                <div class="stat-card-icon">💭</div>
                <div class="stat-card-number">${overall.open_doubts}</div>
                <div class="stat-card-label">Open Doubts</div>
            </div>
        </div>
    `;

    if (data.by_subject && data.by_subject.length > 0) {
        html += `<h3 class="section-title">📚 Performance by Subject</h3>`;
        html += `<div class="subject-progress-grid">`;
        data.by_subject.forEach(s => {
            const icon = s.subject === 'Mathematics' ? '📐' :
                         s.subject === 'Science' ? '🔬' :
                         s.subject === 'English' ? '📖' : '🌍';
            html += `
                <div class="subject-progress-card">
                    <h4>${icon} ${s.subject}</h4>
                    <div class="progress-bar-container">
                        <div class="progress-bar" style="width: ${s.accuracy}%; background: linear-gradient(90deg, var(--primary), var(--accent-pink))"></div>
                    </div>
                    <div class="subject-stats">
                        <span>📊 ${s.accuracy}% accuracy</span>
                        <span>📝 ${s.quizzes} quizzes</span>
                        <span>✅ ${s.correct}/${s.questions}</span>
                    </div>
                </div>
            `;
        });
        html += `</div>`;
    }

    if (data.recent_attempts && data.recent_attempts.length > 0) {
        html += `<h3 class="section-title">📋 Recent Quiz Attempts</h3>`;
        html += `<div class="attempts-list">`;
        data.recent_attempts.forEach(a => {
            const passed = a.percentage >= 60;
            const statusIcon = passed ? '✅' : '⚠️';
            const date = a.date ? new Date(a.date).toLocaleDateString() : 'Pending';
            html += `
                <div class="attempt-item">
                    <div>
                        <h4>${statusIcon} ${a.quiz_title}</h4>
                        <p>📚 ${a.subject} • 📅 ${date}</p>
                    </div>
                    <div class="attempt-score ${passed ? 'pass' : 'fail'}">
                        ${a.percentage ? a.percentage.toFixed(1) + '%' : 'Pending'}
                    </div>
                </div>
            `;
        });
        html += `</div>`;
    } else {
        html += `<p class="info-tip">📝 Your child hasn't taken any quizzes yet. Encourage them to start!</p>`;
    }

    html += `
        <div class="parent-actions">
            <button onclick="viewChildDoubts()" class="btn btn-primary">💭 View Doubts</button>
            <button onclick="searchChild()" class="btn btn-secondary">🔄 Refresh</button>
        </div>
    `;

    container.innerHTML = html;
}

async function viewChildDoubts() {
    if (!currentChildEmail) return;

    try {
        const doubts = await API.request('GET', `/parent/child/${encodeURIComponent(currentChildEmail)}/doubts`);

        let html = `<h2>💭 ${currentChildEmail}'s Doubts</h2>`;
        if (doubts.length === 0) {
            html += `<p>No doubts saved yet.</p>`;
        } else {
            html += `<div class="doubts-list-parent">`;
            doubts.forEach(d => {
                const statusClass = d.status === 'resolved' ? 'resolved' : 'open';
                html += `
                    <div class="doubt-item-parent ${statusClass}">
                        <h4>${d.question}</h4>
                        <p>📚 ${d.subject} • Status: ${d.status} • Responses: ${d.response_count}</p>
                        <p class="date-small">${new Date(d.created_at).toLocaleDateString()}</p>
                    </div>
                `;
            });
            html += `</div>`;
        }
        html += `<button onclick="closeModal()" class="btn btn-primary">Close</button>`;

        openModal(html);
    } catch (error) {
        Utils.showSuccess('Could not load doubts');
    }
}

document.addEventListener('DOMContentLoaded', () => {
    const parentBtn = document.getElementById('parentBtn');
    if (parentBtn) {
        parentBtn.addEventListener('click', () => {
            switchPanel('parentPanel', parentBtn);
            loadParentDashboard();
        });
    }

    // Parent Dashboard button in navbar (for parent role)
    const parentDashBtn = document.getElementById('parentDashBtn');
    if (parentDashBtn) {
        parentDashBtn.addEventListener('click', () => {
            switchPanel('parentPanel', parentDashBtn);
            loadParentDashboard();
        });
    }
});
